# Bài tập 6

Đặng Hồng Minh - 1712096

## Câu 1 (1 điểm)

Định nghĩa của deterministic noise (slide của Thầy Yaser, lecture 11 - overfitting, trang 16): "The part of $f$ that $\mathcal H$ cannot capture: $f(\mathbf{x}) − h^∗(\mathbf{x})$". Trong đó, $h^*(\mathbf{x})$ là hàm tốt nhất của $\mathcal H$ dùng để xấp xỉ $f$.

Ta thấy rằng $\mathcal{H}' \subset \mathcal{H}$, cho nên hàm $h^∗(\mathbf{x})$ của $\mathcal H'$ sẽ không tốt bằng $\mathcal H$. Mà phần của $f$ mà $\mathcal H'$ cannot capture: $f(\mathbf{x}) − h^∗(\mathbf{x})$ sẽ lớn hơn so với $\mathcal H$.

"Deterministic noise" của $\mathcal H'$ sẽ tăng so với $\mathcal H$

<font color = red>**Vậy em chọn câu $[b]\text{ In general, deterministic noise will increase.}$**</font>

## Câu 2 (1 điểm)

In [1]:
import numpy as np

In [2]:
#Load train data
train_data = np.loadtxt("in.dta")
m, n = train_data.shape
train_X = train_data[:,0:2].reshape(m, 2)
train_Y = train_data[:,2].reshape(m, 1)

#Load test data
test_data = np.loadtxt("out.dta")
m, n = test_data.shape

test_X = test_data[:,0:2].reshape(m, 2)
test_Y = test_data[:,2].reshape(m, 1)

In [3]:
#Hàm biến đổi non-linear dữ liệu sang không gian Z
def nonLinearTransform(X):
    N = X.shape[0]
    X_2 = X*X
    X1X2 = np.prod(X, axis = 1).reshape(N,1)
    absSub = np.abs(np.subtract(X[:,0], X[:, 1])).reshape(N, 1)
    absAdd = np.abs(np.add(X[:, 0], X[:, 1])).reshape(N, 1)
    X_nonlinear = np.hstack((np.ones((N, 1)), X, X_2, X1X2, absSub, absAdd)) 
    return X_nonlinear

Z = nonLinearTransform(train_X) # sử dụng để train và tính E_in
Z_test = nonLinearTransform(test_X) # Sử dụng để kiểm tra E_out

In [4]:
#Hàm tính w regularization
def calculate_w_reg(Z, y, lamda):
    shape= Z.shape
    A = np.linalg.inv(np.dot(Z.T, Z) + lamda*np.eye(shape[1]))
    B = np.dot(Z.T, y)
    w_reg = np.dot(A, B)
    return w_reg

In [5]:
#Hàm linear_regression với tham số lamda
def linear_regression(lamda, mode = 0):
    w_reg = calculate_w_reg(Z, train_Y, lamda) # Tính w regularization theo công thức bằng hàm calculate_w_reg
    test_err_out = np.mean(np.sign(np.dot(Z_test,w_reg)) != test_Y) # Tính độ lỗi nhị phân trên tập test E_out 
    test_err_in = np.mean(np.sign(np.dot(Z,w_reg)) != train_Y) # Tính độ lỗi nhị phân trên tập train E_in
    if mode == 0:
        print("E_in = ",test_err_in)
        print("E_out =", test_err_out)
    else:
        return test_err_in, test_err_out

In [6]:
linear_regression(0)

E_in =  0.02857142857142857
E_out = 0.084


$E_{in} \approx 0.03$ và $E_{out} \approx 0.08$

<font color = red>**Vậy em chọn câu $[a]\space 0.03, 0.08$**</font>

## Câu 3 (1 điểm)

Với $k=-3 \Rightarrow$ $\lambda = 0.001$

In [7]:
linear_regression(0.001)

E_in =  0.02857142857142857
E_out = 0.08


$E_{in} \approx 0.03$ và $E_{out} \approx 0.08$

<font color = red>**Vậy em chọn câu $[d]\space 0.03, 0.08$**</font>

## Câu 4 (1 điểm)

Với $k=3 \Rightarrow \lambda=1000$

In [8]:
linear_regression(1000)

E_in =  0.37142857142857144
E_out = 0.436


$E_{in} \approx 0.4$ và $E_{out} \approx 0.4$

<font color = red>**Vậy em chọn câu $[e]\space 0.4, 0.4$**</font>

## Câu 5 (1 điểm)

In [9]:
res = []
#k = 2, lamda = 100
res = linear_regression(100,1)
print("k = 2, lamda = 100 => E_out: ",res[1])

#k = 1, lamda = 10
res = linear_regression(10,1)
print("k = 1, lamda = 10 => E_out: ",res[1])

#k = 0, lamda = 1
res = linear_regression(1,1)
print("k = 0, lamda = 1 => E_out: ",res[1])

#k = -1, lamda = 0.1
res = linear_regression(0.1,1)
print("k = -1, lamda = 0.1 => E_out: ",res[1])

#k = -2, lamda = 0.01
res = linear_regression(0.01,1)
print("k = -2, lamda = 0.01 => E_out: ",res[1])

k = 2, lamda = 100 => E_out:  0.228
k = 1, lamda = 10 => E_out:  0.124
k = 0, lamda = 1 => E_out:  0.092
k = -1, lamda = 0.1 => E_out:  0.056
k = -2, lamda = 0.01 => E_out:  0.084


Giá trị $E_{out}$ nhỏ nhất khi $k=-1$

<font color = red>**Vậy em chọn câu $[d] -1$**</font>

## Câu 6 (1 điểm)

In [10]:
import math
#Ham main cau 6
def mainrun_6():
    E_outs = []
    k = -10
    min = [k, 1000]
    while (k < 10):
        E_outs.append(linear_regression(math.pow(10, k), 1)[1])
        if min[1] > E_outs[len(E_outs) - 1]:
            min[1] = E_outs[len(E_outs) - 1]
            min[0] = k
        k += 1
    return min

In [11]:
res = mainrun_6()
print("k min: ",res[0],",E_out min: ",res[1])

k min:  -1 ,E_out min:  0.056


Với $k \in [-10,10]$ ta vẫn nhận được $E_{out} \approx 0.06$

<font color = red>**Vậy em chọn câu $[b]\space 0.06$**</font>

## Câu 7 (1 điểm)

Ta có:

$\mathcal{H}(10, 0, 3) = w_0 + w_1*L_1(x)+w_2*L_2(x)$

$\mathcal{H}(10, 0, 4) = w_0 + w_1*L_1(x)+w_2*L_2(x) + w_3*L_3(x)$

$\mathcal{H}(10, 1, 3) = w_0 + w_1*L_1(x)+w_2*L_2(x) + L_3 + L_4 + L_5 + L_6 + L_7 + L_8 + L_9 + L_{10}$

$\mathcal{H}(10, 1, 4) = w_0 + w_1*L_1(x)+w_2*L_2(x) + w_3*L_3(x) + L_4 + L_5 + L_6 + L_7 + L_8 + L_9 + L_{10}$

$\mathcal{H}(10, 0, 3) \cup \mathcal{H}(10, 0, 4) = w_0 + w_1*L_1(x)+w_2*L_2(x) + w_3*L_3(x) = \mathcal{H}_3$ (không thỏa $\mathcal{H}_4$)

$\mathcal{H}(10, 1, 3) \cup \mathcal{H}(10, 1, 4) = w_0 + w_1*L_1(x)+w_2*L_2(x) + (w_3)*L_3(x) + L_4 + L_5 + L_6 + L_7 + L_8 + L_9 + L_{10} = \mathcal{H}_{10}$ (không thỏa $\mathcal{H}_3$)

$\mathcal{H}(10, 0, 3) \cap \mathcal{H}(10, 0, 4) = w_0 + w_1*L_1(x)+w_2*L_2(x) = \mathcal{H}_2$ (thỏa)

$\mathcal{H}(10, 1, 3) \cap \mathcal{H}(10, 1, 4) = w_0 + w_1*L_1(x)+w_2*L_2(x) + L_3(x) + L_4 + L_5 + L_6 + L_7 + L_8 + L_9 + L_{10} = \mathcal{H}_{10}$
(không thỏa $\mathcal{H}_1$)

<font color = red>**Vậy em chọn câu $[c]\space\mathcal{H}(10, 0, 3) \cap \mathcal{H}(10, 0, 4) = w_0 + w_1*L_1(x)+w_2*L_2(x) = \mathcal{H}_2$**</font>

## Câu 8 (1 điểm)

1. Bắt đầu từ véc-tơ input $\mathbf{x}$, thực hiện lan truyền tiến qua các tầng của mạng để tính ra các giá trị đầu ra $a$ của các nơ-ron. Quá trình forward số phép tính $w^{(l)}_{ij}a^{(l-1)}_i$: 6 * 3 + 4 = 22 (Vì ở tầng 0 có 6 nút tính luôn cả (+1) fully-connected với tầng 1 là 6 * 3, tầng 1 có 4 nút tính luôn cả (+1) fully-connected với tầng 2 là 4 * 1).

2. Từ $\delta$ ở tầng cuối, thực hiện lan truyền ngược để tính $\delta$ của các nơ-ron ở các tầng trước đó. 

    Tính $\delta^{(1)}$, vì tầng 2 là tầng output chỉ có 1 nút:
    + $\delta_0^{(1)} = \delta_1^{(2)}*w_{01}^{(2)}$
    + $\delta_1^{(1)} = \delta_1^{(2)}*w_{11}^{(2)}$
    + $\delta_2^{(1)} = \delta_1^{(2)}*w_{21}^{(2)}$
    + $\delta_3^{(1)} = \delta_1^{(2)}*w_{31}^{(2)}$
    Suy ra: Có 4 phép tính $w^{(l)}_{ij}\delta^{(l)}_j$ ở bước này. Nhưng không cần tính $\delta_0^{(1)}$, vì nó không được dùng để tính tiếp nên số phép tính cần thiết là 3

    Không cần tính $\delta^{(0)}$ vì nó ở cuối rồi tính chỉ phí.

3. Với mỗi trọng số $w^{(l)}_{ij}$, tính $\frac{\partial e}{\partial w^{(l)}_{ij}}$ bằng cách lấy $a$ ở một đầu (nơ-ron $i$ ở tầng $l-1$) nhân với $\delta$ ở đầu kia (nơ-ron $j$ ở tầng $l$).
    + Xét tầng 2: $\frac{\partial e}{\partial w^{(2)}_{ij}}$ Số phép tính $\delta_j^{(2)}*a_i^{(1)}$ là 4 * 1 = 4
    + Xét tầng 1: $\frac{\partial e}{\partial w^{(1)}_{ij}}$ Số phép tính $\delta_j^{(1)}*a_i^{(0)}$ là 6 * 3 = 18
    
Tổng là 4 + 18 = 22

Vậy tổng số phép tính cần là: 22 + 3 + 22 = 47 gần với đáp án 45 nhất.

<font color = red>**Vậy em chọn câu $[d]\space 45$**</font>

## Câu 9 (1 điểm)

Với 36 nút ẩn ta chia thành 18 tầng ẩn (1 -> 18), vì cần lấy số lượng weights nhỏ nhất nên số weight giữa các tầng ẩn chỉ cần 2 là thỏa yêu cầu.
- Tầng 0: 10 unit (input) -> tầng ẩn 1 1 unit (unit $x^{(1)}_1$): cần 10 weights
- Từ tầng ẩn 1 -> 2, ..., 17 -> 18 ta chỉ cần 2 weights (2 nút ở tầng ẩn trước đến tầng ẩn sau)-> 2 * 17 = 34 weights
- Từ tầng ẩn 18 -> tầng output: 2 weights

Vậy cần tổng là 10 + 34  + 2= 46 weights

<font color = red>**Vậy em chọn câu $[a]\space 46$**</font>

## Câu 10 (1 điểm)

In [12]:
def maxnumWeightsof_2():
    nW = []
    for i in range(2, 35):
        nW.append(10 * (i - 1) + i * (35 - i) + (36 - i))
    return np.max(nW)

In [13]:
def maxnumWeightsof_3():
    nW = []
    for i in range(2, 33):
        for j in range(2, 33):
            if (36 - i - j >= 2):
                nW.append(10*(i-1) + i*(j - 1) + j * (35 - i - j) + (36 - i - j))
            else:
                break
    return np.max(nW)

In [14]:
def maxnumWeightsof_4():
    nW = []
    for i in range(2, 31):
        for i1 in range(2, 31):
            for i2 in range(2, 31):
                if (36 - i - i1- i2 >= 2):
                    nW.append(10*(i-1) + i*(i1 - 1) + i1 * (i2 - 1) + i2 * (35 - i - i1 - i2) + (36 - i - i1 -i2))
                else:
                    break
    return np.max(nW)

In [15]:
def maxnumWeightsof_5():
    nW = []
    for i in range(2, 29):
        for i1 in range(2, 29):
            for i2 in range(2, 29):
                for i3 in range(2, 29):
                    if (36 - i - i1 - i2 - i3 >= 2):
                        nW.append(10*(i-1) + i*(i1 - 1) + i1 * (i2 - 1) + i2*(i3 - 1) + i3 * (35 - i - i1 - i2 - i3) 
                                  + (36 - i - i1 -i2 - i3))
                    else:
                        break
    return np.max(nW)

In [16]:
def maxnumWeightsof_6():
    nW = []
    for i in range(2, 27):
        for i1 in range(2, 27):
            for i2 in range(2, 27):
                for i3 in range(2, 27):
                    for i4 in range(2, 27):
                        if (36 - i - i1 - i2 - i3 -i4 >= 2):
                            nW.append(10*(i-1) + i*(i1 - 1) + i1 * (i2 - 1) + i2*(i3 - 1) 
                                      + i3 * (i4 - 1) + i4 * (35 - i - i1 - i2 - i3 - i4) + (36 - i - i1 -i2 - i3 - i4))
                        else:
                            break
    return np.max(nW)

In [17]:
def maxnumWeightsof_7():
    nW = []
    for i in range(2, 25):
        for i1 in range(2, 25):
            for i2 in range(2, 25):
                for i3 in range(2, 25):
                    for i4 in range(2, 25):
                        for i5 in range(2, 25):
                            if (36 - i - i1 - i2 - i3 - i4 - i5 >= 2):
                                nW.append(10*(i-1) + i*(i1 - 1) + i1 * (i2 - 1) + i2*(i3 - 1) 
                                      + i3 * (i4 - 1) + i4 * (i5 - 1) + i5 * (35 - i - i1 - i2 - i3 - i4 - i5) 
                                      + (36 - i - i1 -i2 - i3 - i4 - i5))
                            else:
                                break
    return np.max(nW)

In [18]:
print("Maximum num of weights with 2 hidden layers: ",maxnumWeightsof_2())
print("Maximum num of weights with 3 hidden layers: ",maxnumWeightsof_3())
print("Maximum num of weights with 4 hidden layers: ",maxnumWeightsof_4())
print("Maximum num of weights with 5 hidden layers: ",maxnumWeightsof_5())
print("Maximum num of weights with 6 hidden layers: ",maxnumWeightsof_6())
print("Maximum num of weights with 7 hidden layers: ",maxnumWeightsof_7())

Maximum num of weights with 2 hidden layers:  510
Maximum num of weights with 3 hidden layers:  467
Maximum num of weights with 4 hidden layers:  426
Maximum num of weights with 5 hidden layers:  387
Maximum num of weights with 6 hidden layers:  350
Maximum num of weights with 7 hidden layers:  315


- Xét mạng có 1 tầng ẩn, số weights là: 10 * 35 + 36 = 386

- Xét mạng có 2 tầng ẩn, số weights lớn nhất là: 10 * (22 - 1) + 22 * (14 - 1) + 14 = 510

- Xét mạng từ 3 tầng ẩn trở lên, khi đó số nút ẩn ko có weight (+1) sẽ tăng theo nên -> số lượng weight sẽ không tăng lên hơn so với 2 tầng ẩn (như kết quả ở trên số tầng ẩn càng tăng thì số weight càng giảm)

$\Rightarrow$ Vậy số weight lớn nhất mà mạng có thể có là 510 với 2 tầng ẩn

<font color = red>**Vậy em chọn câu $[e]\space 510$**</font>